example/lib/sensor_tag_test_scenario.dart

```diff
@@ -79,6 +95,118 @@ class TestScenario {
+        .then((_) {
+          log("Turn off temperature update");
+          return peripheral.writeCharacteristic(
+              SensorTagTemperatureUuids.temperatureService,
+              SensorTagTemperatureUuids.temperatureConfigCharacteristic,
+              Uint8List.fromList([0]),
+              false);
+        })
+        .then((_) {
+          return peripheral.readCharacteristic(
+              SensorTagTemperatureUuids.temperatureService,
+              SensorTagTemperatureUuids.temperatureDataCharacteristic);
+        })
+        .then((data) {
+          log("Temperature value ${data.value}");
+        })
+        .then((_) {
+          log("Turn on temperature update");
+          return peripheral.writeCharacteristic(
+              SensorTagTemperatureUuids.temperatureService,
+              SensorTagTemperatureUuids.temperatureConfigCharacteristic,
+              Uint8List.fromList([1]),
+              false);
+        })
+        .then((_) => Future.delayed(Duration(seconds: 1)))
+        .then((_) {
+          return peripheral.readCharacteristic(
+              SensorTagTemperatureUuids.temperatureService,
+              SensorTagTemperatureUuids.temperatureDataCharacteristic);
+        })
+        .then((data) {
+          log("Temperature value ${data.value}");
+        })
+        .then((_) => log("Test read/write characteristic on service"))
+        .then((_) async {
+          log("Turn off temperature update");
+          Service service = await peripheral.services().then((services) =>
+              services.firstWhere((service) =>
+                  service.uuid ==
+                  SensorTagTemperatureUuids.temperatureService.toLowerCase()));
+          await service.writeCharacteristic(
+            SensorTagTemperatureUuids.temperatureConfigCharacteristic,
+            Uint8List.fromList([0]),
+            false,
+          );
+          return service;
+        })
+        .then((service) =>
+            Future.delayed(Duration(seconds: 1)).then((_) => service))
+        .then((service) async {
+          CharacteristicWithValue dataCharacteristic =
+              await service.readCharacteristic(
+                  SensorTagTemperatureUuids.temperatureDataCharacteristic);
+          return Pair(service, dataCharacteristic);
+        })
+        .then((serviceAndCharacteristic) {
+          log("Temperature value ${serviceAndCharacteristic.second.value}");
+          return serviceAndCharacteristic.first;
+        })
+        .then((service) async {
+          log("Turn on temperature update");
+          Characteristic configCharacteristic =
+              await service.writeCharacteristic(
+                  SensorTagTemperatureUuids.temperatureConfigCharacteristic,
+                  Uint8List.fromList([1]),
+                  false);
+          return Pair(service, configCharacteristic);
+        })
+        .then((serviceAndConfigCharacteristic) =>
```

Comment:
This flow can be simplified by using more of async await. We don't need to chain all of those calls, but the example should read nicely and I don't think it does now. Perhaps some operations should be extracted to separate functions?

Reply:
I created issue to refactor this test case.